File: py/graph/bf_search.py
from collections import deque
from typing import Dict, List
import undirected
import directed


def bf_search(graph: Dict[str, List[str]], start: str, target: str) -> List[str] or None:
    if start == target:
        return [start]

    queue = deque([start])
    parent = {start: None}

    while queue:
        current = queue.popleft()

        for neighbor in graph[current]:
            if neighbor not in parent:
                parent[neighbor] = current

                if neighbor == target:
                    path = []
                    while neighbor is not None:
                        path.append(neighbor)
                        neighbor = parent[neighbor]
                    return path[::-1]

                queue.append(neighbor)

    return None


print(bf_search(directed.perks_graph, directed.ENTRY, directed.ADA))
print(bf_search(undirected.cities_graph, undirected.DE, undirected.SE))

File: ts/src/graph/BFSearch.ts
import * as undirected from './undirected'
import * as directed from './directed'
import { Queue } from '../helpers/queue'
import { arrayNotEmpty } from '../helpers/arrayNotEmpty'

type Graph = Record<string, string[]>

export const BFSearch = (graph: Graph, start: string, target: string): string[] | null => {
	if (start === target) return [start]

	const queue = new Queue([start])
	const parents: Record<string, string | null> = { [start]: null }

	while (arrayNotEmpty(queue.list)) {
		const current = queue.pop()
		const neighbors = graph[current]

		for (let neighbor of neighbors) {
			if (!(neighbor in parents)) {
				parents[neighbor] = current

				if (neighbor === target) {
					let parent: string | null = neighbor
					const path = []

					while (parent) {
						path.push(parent)
						parent = parents[parent]
					}

					return path.reverse()
				} else {
					queue.append(neighbor)
				}
			}
		}
	}

	return null
}

console.log(BFSearch(undirected.citiesGraph, undirected.SE, undirected.BT))
console.log(BFSearch(directed.perksGraph, directed.ENTRY, directed.FAT))
